migrator.py

```diff
@@ -3,39 +3,132 @@
+def migrate_directories_under(input_root_dir: Path, 
+                              db: DigikamDb, 
+                              dry_run: bool,
+                              contacts_file: Optional[Path],
+                              skip_same_rect: Optional[bool]):
     """Traverses directory tree to find directories to migrate."""
+
+    # Build a contact_id to name(s) dictionary
+    global_names: GlobalNames = {}
+
+    if contacts_file is not None:
+        # contacts.xml seems more authoritative than *.ini files
+        # Use this whenever possible.
+        logging.debug(f'Reading contacts from {contacts_file}')
+        tree = ET.parse(contacts_file)
+        for contact in tree.getroot():
+            logging.debug(f'{contact.attrib}')
+            global_names[contact.attrib['id']] = contact.attrib['name']
+
+    else:
+        # Alternate way using *.ini files
+        # Sometimes names will be missing or conflicting if found 
+        # different in multiple directories.  Conflicting names
+        # will be merged by this script by concatenating them with " | " 
+        # separators in sorted order.  
+        contact_id_2_nameset: ContactID2NamesSet = {}
+        for input_dir, subdirs, files in os.walk(input_root_dir):
+            for ini_file in (_PICASA_INI_FILE,_OLD_PICASA_INI_FILE):
+                learn_contact_ids(input_dir,ini_file,contact_id_2_nameset)
+
+        global_names = {k: "|".join(sorted(v)) for k,v in contact_id_2_nameset.items()}
+
+        logging.debug(f"global_names={global_names}")
+
     contact_tags_per_dir: Dict[Path, ContactTags] = {}
+
     for input_dir, subdirs, files in os.walk(input_root_dir):
         dir = Path(input_dir)
         if _PICASA_INI_FILE in files:
-           contact_tags_per_dir[dir] = migrate_directory(dir, files, db, contact_tags_per_dir, dry_run=dry_run)
-        elif any([_is_photo_file(file) for file in files]):
-            logging.warning('Found photos but no %s in %s' % (_PICASA_INI_FILE, dir))
-
+            ini_file = _PICASA_INI_FILE
+        elif _OLD_PICASA_INI_FILE in files:
+            ini_file = _OLD_PICASA_INI_FILE
+        else:
+            if any([_is_photo_file(file) for file in files]):
+                logging.warning(f'Found photos but no .ini in {dir}')
+            else:
+                logging.warning(f"No photos and no .ini file in {dir}")
+            continue
+        logging.debug(f"Processing {Path(dir/ini_file)}")
+        try:
+            contact_tags_per_dir[dir] = migrate_directory(dir, files, db, 
+                                        contact_tags_per_dir, global_names,
+                                        dry_run=dry_run, ini_file_name=ini_file,
+                                        contacts_file=contacts_file,
+                                        skip_same_rect=skip_same_rect)
+        except Exception as e:
+            # Seems to happen with very old "Originals" directories. digiKam Windows version
```

Comment:
It catches this.  How would you like to resolve it?

```
INFO: Now migrating d:\My Documents\Photos\Xmas 2007 (KW)\Originals
WARNING: Exception: No digiKam Album found for d:\My Documents\Photos\Xmas 2007 (KW)\Originals (relative path /Xmas 2007 (KW)/Originals) under root 1
WARNING: Traceback (most recent call last):
  File "D:\My Documents\Photos\picasa2digikam\migrator.py", line 108, in migrate_directories_under
    contact_tags_per_dir[dir] = migrate_directory(dir, files, db,
  File "D:\My Documents\Photos\picasa2digikam\migrator.py", line 134, in migrate_directory
    album_id = db.find_album_by_dir(input_dir)
  File "D:\My Documents\Photos\picasa2digikam\digikam_db.py", line 99, in find_album_by_dir
    raise ValueError('No digiKam Album found for %s (relative path %s) under root %s' % (path, relative_path, root_id))
ValueError: No digiKam Album found for d:\My Documents\Photos\Xmas 2007 (KW)\Originals (relative path /Xmas 2007 (KW)/Originals) under root 1

```

Reply:
Can you use a tool like DB Browser for SQLite to inspect your database, do sth like `SELECT * FROM Albums WHERE albumRoot = 1` and then look for `Xmas 2007` in the output? I wonder if the album simply doesn't exist (if so, why didn't digiKam create it? Is it excluded explicitly, does it not contain any photos, is there a digiKam bug?) or if it just didn't get matched, perhaps because those parentheses on the path need to be escaped or so.

If there's a systematic issue, it's worth fixing (e.g. we wouldn't want to silently drop and not migrate all photos in directories whose path contains a special character like a parenthesis). If it's a very rare case, the user should make the decision explicitly on a per-directory basis, e.g. a flag `--skip_dir="Xmas 2007 (KW)"`, or even just removing the directory before running the tool, or adding a special marker file `picasa2digikam.ignore` into the directory to have the tool skip it.

Reply:
It wasn't documented anywhere but digiKam will not put anything underneath "Originals" folders in its database.  This affects Windows digiKam but not Linux.

I think you just want to return cleanly from find_album_by_dir() when album_id is None.  How would you like to do that?

Reply:
> digiKam will not put anything underneath "Originals" folders in its database.

And is that a desirable behavior?

The [idea](https://www.digikam.org/news/2010-06-17_gsoc_explaining_digikam_s_non-destructive_editing_concept/) seems to be that one puts the original files in there for backup, but doesn't want to view them in digiKam (because they'd be duplicated, and the backup shouldn't be touched, so not having it visible in digiKam reduces the risk of accidental corruption).

Does this philosophy apply to the "Originals" folder in your use case? Or did you make this for a different reason and you expect it to show up in Picasa/digiKam?

-------

> return cleanly from find_album_by_dir() when album_id is None. How would you like to do that?

Yes. Let's return `Optional[int]` from the function (remove the `if album_id is None` there, just `return self._fetchcell(...)`), and raise the exception at the call site, unless the directory name is `"Originals"`:

```
album_id = db.find_album_by_dir(input_dir)
if album is None:
  if input_dir.name == "Originals"
    logging.warning(...)
    return {}
  raise ValueError(...)
```

The error message won't contain the root_id and relative path anymore, but that's fine.

Reply:
I think i found the pattern -- all my "Originals" directories had the hidden attribute, so I think that is why digiKam ignored them.

![image](https://user-images.githubusercontent.com/13840403/173687974-425037f5-5ed2-4d8b-95ee-78d9761306e4.png)

In addition digiKam has a built-in ignore list that is not well publicised.  Plus it has user configurable ignore lists!

![image](https://user-images.githubusercontent.com/13840403/173687126-9457d8cc-58ec-49fe-baab-ed66a9d47b4e.png)

Since the list of ignored directories is a moving target, I think it would be cleaner to just log the missing digiKam album and continue.

Reply:
Ok agreed. As it's not like digiKam has `== "Originals"` in its code, we also don't want it here.

Reply:
Now returns None when no album found. And caller checks for the None being returned and gracefully continues.

Reply:
Looks good. So now the `try .. except` in `migrate_directories_under()` can be removed, right? (Which was the origin of this discussion.)